### PR TITLE
fix(git): Refactor Git URL resolution code & remove hash from URL

### PR DESCRIPTION
**Pull request by BYK**
**Summary**

This PR is triggered by https://git.io/vdzI2. It refactors the
code to allow more hosted Git expansions (like GitLab) in a more
flexible way. It also prefers `url.parse` over regular expressions
and removes any hasehs in `GitUrl.repository` results.

**Test plan**

Existing tests should pass along with the hash removal correction.

### CODE REVIEW

File: src/util/git.js [hunk 1]
```diff
@@ -33,6 +37,27 @@ const handleSpawnError = err => {
...
+  }),
+  'bitbucket:': parsedUrl => ({
+    ...parsedUrl,
+    slashes: true,
+    auth: 'git',
+    protocol: SSH_PROTOCOL,
+    host: 'bitbucket.com',
+    hostname: 'bitbucket.com',
+    pathname: `/${parsedUrl.hostname}${parsedUrl.pathname}`,
+  }),
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/util/git.js]: There are 2 questions here so I'll answer them separately :)

1. Why is GitLab one of them?

Because it wasn't there already and this is just a refactor so we can add them easily in a follow-up feature patch :)

2. Why don't we generalize it more?

I thought about a similar pattern but it makes too many assumptions about the structure like what and how `auth` should be or how the pathname should be expanded and your example even abstracts out the hostname part :D We may go there in the future probably but I don't want to overly restrict this part.

As an addendum to this, I considered having custom expansions through `.yarnrc` file, something that would translate into
```
shorthandServices: {
    github: {
        auth: 'git',
        hostname: 'github.com',
    }
}
```

This way we can let people define these themselves externally (even with plugin packages may be?). If we do this in the yarnrc file, it will be more limited, if we allow plugins, it would be a lot more flexible but potentially complex, since plugins. I still think people may find this useful defining internal company shortcuts like SSH host aliases.

TL;DR
Let's keep it a bit more flexible as it is to keep us open to new opportunities?

File: src/util/git.js [hunk 2]
```diff
@@ -60,29 +85,39 @@ export default class Git implements GitRefResolvingInterface {
...
+    ) {
       return {
-        hostname: match[2],
-        protocol: 'ssh:',
-        repository: match[1],
+        hostname: parsed.hostname,
+        protocol: parsed.protocol,
+        repository: `${parsed.auth ? parsed.auth + '@' : ''}${parsed.hostname}:${parsed.path.slice(
+          SCP_PATH_PREFIX.length,
+        )}`,
```
[the remark this answers is not in the document]
Comment (BYK) [reply src/util/git.js]: Yeah, good idea. It was simple when I started and then tests failed due to auth lacking :D

**Comment by BYK** [comment]
@kaylieEB ready for another review

File: src/util/git.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaylieEB) [reply src/util/git.js]: thanks for explaining :)

[the remark this answers is not in the document]
Comment (BYK) [reply src/util/git.js]: Thanks a lot for asking since I was able to do a brain dump. Any thoughts on them?

[the remark this answers is not in the document]
Comment (kaylieEB) [reply src/util/git.js]: I like the idea of putting it in `.yarnrc` -- i'd imagine most people would just need to configure host or hostname (maybe protocol?) and don't need to deal with the overhead of plugins beyond that.
